File: src/main/java/brightspark/landmanager/LandManager.java
package brightspark.landmanager;

import brightspark.landmanager.data.CapStorage;
import brightspark.landmanager.data.CapabilityAreas;
import brightspark.landmanager.data.CapabilityAreasImpl;
import brightspark.landmanager.gui.GuiHandler;
import brightspark.landmanager.item.LMItems;
import brightspark.landmanager.message.MessageCreateArea;
import brightspark.landmanager.message.MessageCreateAreaReply;
import brightspark.landmanager.message.MessageUpdateCapability;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.ItemStack;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityInject;
import net.minecraftforge.common.capabilities.CapabilityManager;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import net.minecraftforge.fml.common.event.FMLServerStartingEvent;
import net.minecraftforge.fml.common.network.NetworkRegistry;
import net.minecraftforge.fml.common.network.simpleimpl.SimpleNetworkWrapper;
import net.minecraftforge.fml.relauncher.Side;
import org.apache.logging.log4j.Logger;

@Mod(modid = LandManager.MOD_ID, name = LandManager.MOD_NAME, version = LandManager.VERSION)
public class LandManager
{
    public static final String MOD_ID = "landmanager";
    public static final String MOD_NAME = "Land Manager";
    public static final String VERSION = "@VERSION@";

    @Mod.Instance(MOD_ID)
    public static LandManager INSTANCE;
    public static Logger LOGGER;
    public static SimpleNetworkWrapper NETWORK;

    @CapabilityInject(CapabilityAreas.class)
    public static Capability<CapabilityAreas> CAPABILITY_AREAS = null;

    public static final CreativeTabs LM_TAB = new CreativeTabs(MOD_ID)
    {
        @Override
        public ItemStack getTabIconItem()
        {
            return new ItemStack(LMItems.adminItem);
        }
    };

    @Mod.EventHandler
    public static void preInit(FMLPreInitializationEvent event)
    {
        LOGGER = event.getModLog();

        NetworkRegistry.INSTANCE.registerGuiHandler(INSTANCE, new GuiHandler());
        NETWORK = NetworkRegistry.INSTANCE.newSimpleChannel(MOD_ID);
        NETWORK.registerMessage(new MessageCreateArea.Handler(), MessageCreateArea.class, 0, Side.SERVER);
        NETWORK.registerMessage(new MessageCreateAreaReply.Handler(), MessageCreateAreaReply.class, 1, Side.CLIENT);
        NETWORK.registerMessage(new MessageUpdateCapability.Handler(), MessageUpdateCapability.class, 2, Side.CLIENT);

        CapabilityManager.INSTANCE.register(CapabilityAreas.class, new CapStorage<>(), CapabilityAreasImpl::new);
    }

    @Mod.EventHandler
    public void serverStarting(FMLServerStartingEvent event)
    {
        //TODO: Commands
    }
}

File: src/main/kotlin/fr/bloctave/lmr/LandManager.kt
package fr.bloctave.lmr

import com.mojang.brigadier.context.CommandContext
import fr.bloctave.lmr.data.requests.ClientRequest
import fr.bloctave.lmr.message.MessageRequestAdd
import fr.bloctave.lmr.message.MessageRequestDelete
import fr.bloctave.lmr.api.create.CreateProxy
import fr.bloctave.lmr.api.effortlessbuilding.EffortlessBuildingProxy
import fr.bloctave.lmr.api.holefiller.HoleFillerProxy
import fr.bloctave.lmr.api.iceandfire.IceAndFireProxy
import fr.bloctave.lmr.api.supplementaries.SupplementariesProxy
import fr.bloctave.lmr.command.LMCommand
import fr.bloctave.lmr.command.argumentType.AreaArgument
import fr.bloctave.lmr.command.argumentType.RequestArgument
import fr.bloctave.lmr.init.LMBlocks
import fr.bloctave.lmr.init.LMCapabilities
import fr.bloctave.lmr.init.LMItems
import fr.bloctave.lmr.message.*
import fr.bloctave.lmr.api.proxy.DependencyProxy

import fr.bloctave.lmr.api.vanilla.VanillaProxy
import fr.bloctave.lmr.config.ClientConfig
import fr.bloctave.lmr.config.CommonConfig
import fr.bloctave.lmr.config.ServerConfig

import fr.bloctave.lmr.util.*
import net.minecraft.command.CommandSource
import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraft.item.ItemGroup
import net.minecraft.item.ItemStack
import net.minecraft.server.MinecraftServer
import net.minecraft.util.ResourceLocation
import net.minecraftforge.event.RegisterCommandsEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.config.ModConfig
import net.minecraftforge.fml.config.ModConfig.ModConfigEvent
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent
import net.minecraftforge.fml.network.NetworkEvent
import net.minecraftforge.fml.network.NetworkRegistry
import net.minecraftforge.fml.network.simple.SimpleChannel
import org.apache.logging.log4j.LogManager
import org.spongepowered.asm.mixin.MixinEnvironment
import thedarkcolour.kotlinforforge.forge.FORGE_BUS
import thedarkcolour.kotlinforforge.forge.MOD_BUS
import thedarkcolour.kotlinforforge.forge.registerConfig


@Mod(LandManager.MOD_ID)
object LandManager {
	const val MOD_ID = "lmr"
	val LOGGER = LogManager.getLogger(LandManager::class.java)

	val group = object : ItemGroup(MOD_ID) {
		override fun makeIcon() = ItemStack(area_create!!)
	}

	private val dependencyProxy = DependencyProxy(MOD_ID)

	private val clientConfig = ClientConfig
	private val commonConfig = CommonConfig
	private val serverConfig = ServerConfig

	private val clientRequest: ClientRequest

	private const val NETWORK_PROTOCOL = "1"
	val NETWORK: SimpleChannel = NetworkRegistry.newSimpleChannel(
		ResourceLocation(MOD_ID, "main"),
		{ NETWORK_PROTOCOL },
		NETWORK_PROTOCOL::equals,
		NETWORK_PROTOCOL::equals
	).apply {
		arrayOf(
			MessageRequestAdd::class,
			MessageRequestDelete::class,
			MessageAreaAdd::class,
			MessageAreaChange::class,
			MessageAreaDelete::class,
			MessageAreaRename::class,
			MessageChatLog::class,
			MessageCreateArea::class,
			MessageCreateAreaReply::class,
			MessageHomeActionAdd::class,
			MessageHomeActionKickOrPass::class,
			MessageHomeActionReply::class,
			MessageHomeActionReplyMessage::class,
			MessageHomeToggle::class,
			MessageHomeToggleReply::class,
			MessageOpenCreateAreaGui::class,
			MessageOpenHomeGui::class,
			MessageShowArea::class,
			MessageUpdateAreasCap::class
		).forEachIndexed { i, kClass -> registerMessage(kClass, i) }
	}

	/*val EVENT_NETWORK_CHANNEL = NetworkRegistry.newEventChannel(
		ResourceLocation(MOD_ID, "listener"),
		{ NETWORK_PROTOCOL },
		NETWORK_PROTOCOL::equals,
		NETWORK_PROTOCOL::equals
	).apply { it -> it.register(this, MessageAreaChange::class.java, 0) }*/

	init {

		MOD_BUS.apply {
			addListener<ModConfigEvent> {
				if (it.config.modId == MOD_ID) {
					/*clientConfig.register()
					commonConfig.register()
					serverConfig.register()*/
				}
			}
			addListener<FMLCommonSetupEvent> {
				it.enqueueWork { LMCapabilities.register() }
			}

			addGenericListener(LMBlocks::register)
			addGenericListener(LMItems::register)
		}
		FORGE_BUS.apply {
			addGenericListener(LMCapabilities::attach)
			addListener(LMCapabilities::playerLoggedIn)
			addListener(LMCapabilities::playerRespawn)
			addListener(LMCapabilities::playerChangedDimension)
			addListener<RegisterCommandsEvent> { it.dispatcher.register(LMCommand) }
		}

		NETWORK_PROTOCOL



		regCommandArgType<AreaArgument>("area")
		regCommandArgType<RequestArgument>("request")


		registerConfig(ModConfig.Type.CLIENT, clientConfig.bakeConfig().second)
		registerConfig(ModConfig.Type.COMMON, commonConfig.bakeConfig().second)
		registerConfig(ModConfig.Type.SERVER, serverConfig.bakeConfig().second)

		MixinEnvironment.Side.CLIENT.run {
			clientRequest = ClientRequest()
		}

		dependencyProxy.addDependency(VanillaProxy)
		dependencyProxy.addDependency(IceAndFireProxy)
		dependencyProxy.addDependency(SupplementariesProxy)
		dependencyProxy.addDependency(HoleFillerProxy)
		dependencyProxy.addDependency(EffortlessBuildingProxy)
		dependencyProxy.addDependency(CreateProxy)

		dependencyProxy.registerDependencies()

		//MyPacketHandler()
		//EVENT_NETWORK_CHANNEL.addListener(this::handleEvent)

	}

	fun areaChange(context: CommandContext<CommandSource>, type: AreaChangeType, areaName: String) =
		areaChange(context.source.server, type, areaName, context.getSenderName())

	fun areaChange(
		server: MinecraftServer,
		type: AreaChangeType,
		areaName: String,
		sender: ServerPlayerEntity? = null
	) {
		val timestamp = System.currentTimeMillis()
		val senderName = sender?.gameProfile?.name ?: server.name()
		server.playerList.ops.userList
			.mapNotNull { server.playerList.getPlayerByName(it) }
			.filter { it != sender }
			.forEach { NETWORK.sendToPlayer(MessageChatLog(timestamp, type, areaName, senderName), it) }
	}

	private fun areaChange(server: MinecraftServer, type: AreaChangeType, areaName: String, name: String) {
		val timestamp = System.currentTimeMillis()
		server.playerList.ops.userList
			.filter { !it.equals(name, true) }
			.mapNotNull { server.playerList.getPlayerByName(it) }
			.forEach { NETWORK.sendToPlayer(MessageChatLog(timestamp, type, areaName, name), it) }
	}

	fun getDependencyProxy(): DependencyProxy = dependencyProxy


	@SubscribeEvent
	fun handleEvent(event: NetworkEvent.ServerCustomPayloadEvent) = event.run {
		LOGGER.info("packet received")
	}

	fun getRequests(): ClientRequest {
		return clientRequest
	}
}
